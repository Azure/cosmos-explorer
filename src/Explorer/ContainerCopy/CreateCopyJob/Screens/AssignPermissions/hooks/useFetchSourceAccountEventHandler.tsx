import { useCallback, useState } from "react";
import { fetchDatabaseAccount } from "../../../../../../Utils/arm/databaseAccountUtils";
import { useCopyJobContext } from "../../../../Context/CopyJobContext";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";

const useFetchSourceAccountEventHandler = () => {
  const [loading, setLoading] = useState<boolean>(false);
  const { copyJobState: { source } = {}, setCopyJobState } = useCopyJobContext();
  const onWindowClosed = useCallback(async (errorMessage: string) => {
    try {
      const selectedSourceAccount = source?.account;
      const {
        subscriptionId: sourceSubscriptionId,
        resourceGroup: sourceResourceGroup,
        accountName: sourceAccountName,
      } = getAccountDetailsFromResourceId(selectedSourceAccount?.id);

      setLoading(true);
      const account = await fetchDatabaseAccount(sourceSubscriptionId, sourceResourceGroup, sourceAccountName);
      if (account) {
        setCopyJobState((prevState) => ({
          ...prevState,
          source: { ...prevState.source, account: account },
        }));
      }
    } catch (error) {
      console.error(errorMessage, error);
    } finally {
      setLoading(false);
    }
  }, []);

  return { loading, onWindowClosed };
};

export default useFetchSourceAccountEventHandler;
