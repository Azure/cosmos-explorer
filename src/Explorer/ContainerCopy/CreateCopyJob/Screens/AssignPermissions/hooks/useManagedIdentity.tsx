import { DatabaseAccount } from "Contracts/DataModels";
import { useCallback, useState } from "react";
import { useCopyJobContext } from "../../../../Context/CopyJobContext";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";

interface UseManagedIdentityUpdaterParams {
    updateIdentityFn: (
        subscriptionId: string,
        resourceGroup?: string,
        accountName?: string
    ) => Promise<DatabaseAccount | undefined>;
}

interface UseManagedIdentityUpdaterReturn {
    loading: boolean;
    handleAddSystemIdentity: () => Promise<void>;
}

const useManagedIdentity = (
    updateIdentityFn: UseManagedIdentityUpdaterParams["updateIdentityFn"]
): UseManagedIdentityUpdaterReturn => {
    const { copyJobState, setCopyJobState } = useCopyJobContext();
    const [loading, setLoading] = useState<boolean>(false);

    const handleAddSystemIdentity = useCallback(async (): Promise<void> => {
        try {
            setLoading(true);
            const selectedTargetAccount = copyJobState?.target?.account;
            const {
                subscriptionId: targetSubscriptionId,
                resourceGroup: targetResourceGroup,
                accountName: targetAccountName
            } = getAccountDetailsFromResourceId(selectedTargetAccount?.id);

            const updatedAccount = await updateIdentityFn(
                targetSubscriptionId,
                targetResourceGroup,
                targetAccountName
            );
            if (updatedAccount) {
                setCopyJobState((prevState) => ({
                    ...prevState,
                    target: { ...prevState.target, account: updatedAccount }
                }));
            }
        } catch (error) {
            console.error("Error enabling system-assigned managed identity:", error);
        } finally {
            setLoading(false);
        }
    }, [copyJobState, updateIdentityFn, setCopyJobState]);

    return { loading, handleAddSystemIdentity };
};

export default useManagedIdentity;