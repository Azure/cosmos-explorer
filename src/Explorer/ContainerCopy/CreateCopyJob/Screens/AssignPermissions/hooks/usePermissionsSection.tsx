import { useEffect, useMemo, useRef, useState } from "react";
import { fetchRoleAssignments, fetchRoleDefinitions, RoleDefinitionType } from "../../../../../../Utils/arm/RbacUtils";
import ContainerCopyMessages from "../../../../ContainerCopyMessages";
import { getAccountDetailsFromResourceId } from "../../../../CopyJobUtils";
import { BackupPolicyType, CopyJobMigrationType, DefaultIdentityType, IdentityType } from "../../../../Enums";
import { CopyJobContextState } from "../../../../Types";
import { useCopyJobPrerequisitesCache } from "../../../Utils/useCopyJobPrerequisitesCache";
import AddManagedIdentity from "../AddManagedIdentity";
import AddReadPermissionToDefaultIdentity from "../AddReadPermissionToDefaultIdentity";
import DefaultManagedIdentity from "../DefaultManagedIdentity";
import OnlineCopyEnabled from "../OnlineCopyEnabled";
import PointInTimeRestore from "../PointInTimeRestore";

export interface PermissionSectionConfig {
  id: string;
  title: string;
  Component: React.ComponentType;
  disabled: boolean;
  completed?: boolean;
  validate?: (state: CopyJobContextState) => boolean | Promise<boolean>;
}

// Section IDs for maintainability
export const SECTION_IDS = {
  addManagedIdentity: "addManagedIdentity",
  defaultManagedIdentity: "defaultManagedIdentity",
  readPermissionAssigned: "readPermissionAssigned",
  pointInTimeRestore: "pointInTimeRestore",
  onlineCopyEnabled: "onlineCopyEnabled",
} as const;

const PERMISSION_SECTIONS_CONFIG: PermissionSectionConfig[] = [
  {
    id: SECTION_IDS.addManagedIdentity,
    title: ContainerCopyMessages.addManagedIdentity.title,
    Component: AddManagedIdentity,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const targetAccountIdentityType = (state?.target?.account?.identity?.type ?? "").toLowerCase();
      return (
        targetAccountIdentityType === IdentityType.SystemAssigned ||
        targetAccountIdentityType === IdentityType.UserAssigned
      );
    },
  },
  {
    id: SECTION_IDS.defaultManagedIdentity,
    title: ContainerCopyMessages.defaultManagedIdentity.title,
    Component: DefaultManagedIdentity,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const targetAccountDefaultIdentity = (state?.target?.account?.properties?.defaultIdentity ?? "").toLowerCase();
      return targetAccountDefaultIdentity === DefaultIdentityType.SystemAssignedIdentity;
    },
  },
  {
    id: SECTION_IDS.readPermissionAssigned,
    title: ContainerCopyMessages.readPermissionAssigned.title,
    Component: AddReadPermissionToDefaultIdentity,
    disabled: true,
    validate: async (state: CopyJobContextState) => {
      const principalId = state?.target?.account?.identity?.principalId;
      const selectedSourceAccount = state?.source?.account;
      const {
        subscriptionId: sourceSubscriptionId,
        resourceGroup: sourceResourceGroup,
        accountName: sourceAccountName,
      } = getAccountDetailsFromResourceId(selectedSourceAccount?.id);

      const rolesAssigned = await fetchRoleAssignments(
        sourceSubscriptionId,
        sourceResourceGroup,
        sourceAccountName,
        principalId,
      );

      const roleDefinitions = await fetchRoleDefinitions(rolesAssigned ?? []);
      return checkTargetHasReaderRoleOnSource(roleDefinitions ?? []);
    },
  },
];

const PERMISSION_SECTIONS_FOR_ONLINE_JOBS: PermissionSectionConfig[] = [
  {
    id: SECTION_IDS.pointInTimeRestore,
    title: ContainerCopyMessages.pointInTimeRestore.title,
    Component: PointInTimeRestore,
    disabled: true,
    validate: (state: CopyJobContextState) => {
      const sourceAccountBackupPolicy = state?.source?.account?.properties?.backupPolicy?.type ?? "";
      return sourceAccountBackupPolicy === BackupPolicyType.Continuous;
    },
  },
  {
    id: SECTION_IDS.onlineCopyEnabled,
    title: ContainerCopyMessages.onlineCopyEnabled.title,
    Component: OnlineCopyEnabled,
    disabled: true,
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    validate: (_state: CopyJobContextState) => {
      return false;
    },
  },
];

/**
 * Checks if the user has the Reader role based on role definitions.
 */
export function checkTargetHasReaderRoleOnSource(roleDefinitions: RoleDefinitionType[]): boolean {
  return roleDefinitions?.some(
    (role) =>
      role.name === "00000000-0000-0000-0000-000000000001" ||
      role.permissions.some(
        (permission) =>
          permission.dataActions.includes("Microsoft.DocumentDB/databaseAccounts/readMetadata") &&
          permission.dataActions.includes("Microsoft.DocumentDB/databaseAccounts/sqlDatabases/containers/items/read"),
      ),
  );
}

/**
 * Returns the permission sections configuration for the Assign Permissions screen.
 * Memoizes derived values for performance and decouples logic for testability.
 */
const usePermissionSections = (state: CopyJobContextState): PermissionSectionConfig[] => {
  const { validationCache, setValidationCache } = useCopyJobPrerequisitesCache();
  const [permissionSections, setPermissionSections] = useState<PermissionSectionConfig[] | null>(null);
  const isValidatingRef = useRef(false);

  const sectionToValidate = useMemo(() => {
    const baseSections = [...PERMISSION_SECTIONS_CONFIG];
    if (state.migrationType === CopyJobMigrationType.Online) {
      return [...baseSections, ...PERMISSION_SECTIONS_FOR_ONLINE_JOBS];
    }
    return baseSections;
  }, [state.migrationType]);

  const memoizedValidationCache = useMemo(() => {
    if (state.migrationType === CopyJobMigrationType.Offline) {
      validationCache.delete(SECTION_IDS.pointInTimeRestore);
      validationCache.delete(SECTION_IDS.onlineCopyEnabled);
    }
    return validationCache;
  }, [state.migrationType]);

  useEffect(() => {
    const validateSections = async () => {
      if (isValidatingRef.current) {
        return;
      }

      isValidatingRef.current = true;
      const result: PermissionSectionConfig[] = [];
      const newValidationCache = new Map(memoizedValidationCache);

      for (let i = 0; i < sectionToValidate.length; i++) {
        const section = sectionToValidate[i];

        // Check if this section was already validated and passed
        if (newValidationCache.has(section.id) && newValidationCache.get(section.id) === true) {
          result.push({ ...section, completed: true });
          continue;
        }
        // We've reached the first non-cached section - validate it
        if (section.validate) {
          const isValid = await section.validate(state);
          newValidationCache.set(section.id, isValid);
          result.push({ ...section, completed: isValid });
          // Stop validation if current section failed
          if (!isValid) {
            for (let j = i + 1; j < sectionToValidate.length; j++) {
              result.push({ ...sectionToValidate[j], completed: false });
            }
            break;
          }
        } else {
          // Section has no validate method
          newValidationCache.set(section.id, false);
          result.push({ ...section, completed: false });
        }
      }

      setValidationCache(newValidationCache);
      setPermissionSections(result);
      isValidatingRef.current = false;
    };

    validateSections();

    return () => {
      isValidatingRef.current = false;
    };
  }, [state, sectionToValidate]);

  return permissionSections ?? [];
};

export default usePermissionSections;
