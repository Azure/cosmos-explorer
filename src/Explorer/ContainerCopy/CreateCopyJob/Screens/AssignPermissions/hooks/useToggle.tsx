import { useCallback, useState } from "react";

const useToggle = (initialState = false) => {
  const [state, setState] = useState<boolean>(initialState);
  const onToggle = useCallback((_, checked?: boolean) => {
    setState(checked);
  }, []);
  return [state, onToggle] as const;
};

export default useToggle;
