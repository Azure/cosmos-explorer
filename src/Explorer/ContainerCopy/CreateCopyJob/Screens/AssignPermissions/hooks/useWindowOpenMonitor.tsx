import { useEffect, useRef } from "react";

const useWindowOpenMonitor = (url: string, onClose?: () => void, intervalMs = 500) => {
    const intervalRef = useRef<NodeJS.Timeout | null>(null);

    const openWindowAndMonitor = () => {
        const newWindow = window.open(url, '_blank');
        intervalRef.current = setInterval(() => {
            if (newWindow?.closed) {
                clearInterval(intervalRef.current!);
                intervalRef.current = null;
                console.log('New window has been closed!');
                if (onClose) {
                    onClose();
                }
            }
        }, intervalMs);
    };

    useEffect(() => {
        return () => {
            if (intervalRef.current) {
                clearInterval(intervalRef.current);
                intervalRef.current = null;
            }
        };
    }, []);

    return openWindowAndMonitor;
};

export default useWindowOpenMonitor;