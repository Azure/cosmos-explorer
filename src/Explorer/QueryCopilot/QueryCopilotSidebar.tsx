import { Icon, IconButton, Image, Stack, Text, TextField } from "@fluentui/react";
import Explorer from "Explorer/Explorer";
import { WelcomeSidebarPopup } from "Explorer/QueryCopilot/Popup/WelcomeSidebarPopup";
import { useQueryCopilotSidebar } from "hooks/useQueryCopilotSidebar";
import React from "react";
import CopilotIcon from "../../../images/CopilotSidebarLogo.svg";

interface QueryCopilotSidebarProps {
  explorer?: Explorer;
}

export const QueryCopilotSidebar: React.FC<QueryCopilotSidebarProps> = ({}: QueryCopilotSidebarProps): JSX.Element => {
  const {
    setWasCopilotUsed,
    showCopilotSidebar,
    setShowCopilotSidebar,
    userInput,
    setUserInput,
    chatMessages,
    setChatMessages,
    showWelcomeSidebar,
  } = useQueryCopilotSidebar();

  const handleSendMessage = () => {
    if (userInput.trim() !== "") {
      setChatMessages([...chatMessages, userInput]);
      setUserInput("");
    }
  };

  const handleInputChange = (value: string) => {
    setUserInput(value);
  };

  const handleEnterKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter" && !event.shiftKey) {
      event.preventDefault();
      handleSendMessage();
    }
  };

  React.useEffect(() => {
    if (showCopilotSidebar) {
      setWasCopilotUsed(true);
    }
  }, []);

  return (
    <Stack style={{ width: "320px", height: "405px", backgroundColor: "#FAFAFA", overflow: "auto" }}>
      {showWelcomeSidebar ? (
        <Stack.Item styles={{ root: { textAlign: "center", verticalAlign: "middle" } }}>
          <WelcomeSidebarPopup />
        </Stack.Item>
      ) : (
        <>
          <Stack style={{ margin: "15px 0px 0px 0px", padding: "5px" }}>
            <Stack style={{ display: "flex", justifyContent: "space-between" }} horizontal verticalAlign="center">
              <Stack horizontal verticalAlign="center">
                <Image src={CopilotIcon} />
                <Text style={{ marginLeft: "5px", fontWeight: "bold" }}>Copilot</Text>
                <Text
                  style={{
                    background: "#f0f0f0",
                    fontSize: "10px",
                    padding: "2px 4px",
                    marginLeft: "5px",
                    borderRadius: "8px",
                  }}
                >
                  Preview
                </Text>
              </Stack>
              <IconButton
                onClick={() => setShowCopilotSidebar(false)}
                iconProps={{ iconName: "Cancel" }}
                title="Exit"
                ariaLabel="Exit"
                style={{ color: "#424242", verticalAlign: "middle" }}
              />
            </Stack>
          </Stack>

          <Stack horizontalAlign="center" tokens={{ padding: 8, childrenGap: 8 }}>
            {new Date().toLocaleDateString("en-US", {
              month: "long",
              day: "numeric",
            })}{" "}
            {new Date().toLocaleTimeString("en-US", {
              hour: "numeric",
              minute: "numeric",
              hour12: true,
            })}
          </Stack>
          <Stack style={{ flexGrow: 1, display: "flex", flexDirection: "column" }}>
            <div style={{ flexGrow: 1, overflowY: "auto" }}>
              <Stack
                tokens={{ padding: 16, childrenGap: 12 }}
                style={{
                  backgroundColor: "white",
                  margin: "5px 10px",
                  borderRadius: "8px",
                }}
              >
                <Text variant="medium">
                  Hello, I am Cosmos Db&apos;s copilot assistant. I can help you do the following things:
                </Text>
                <Stack tokens={{ childrenGap: 8 }}>
                  <Stack horizontal tokens={{ childrenGap: 8 }}>
                    <Icon iconName="BulletedList" />
                    <Text>Generate queries based upon prompt you suggest</Text>
                  </Stack>
                  <Stack horizontal tokens={{ childrenGap: 8 }}>
                    <Icon iconName="BulletedList" />
                    <Text>Explain and provide alternate queries for a query suggested by you</Text>
                  </Stack>
                  <Stack horizontal tokens={{ childrenGap: 8 }}>
                    <Icon iconName="BulletedList" />
                    <Text>Help answer questions about Cosmos DB</Text>
                  </Stack>
                </Stack>
                <Text variant="medium">
                  To get started, ask me a question or use one of the sample prompts to generate a query. AI-generated
                  content may be incorrect.
                </Text>
              </Stack>
              {chatMessages.map((message, index) => (
                <Stack
                  key={index}
                  horizontalAlign="center"
                  tokens={{ padding: 8, childrenGap: 8 }}
                  style={{
                    backgroundColor: "#E0E7FF",
                    borderRadius: "8px",
                    margin: "5px 10px",
                    textAlign: "start",
                  }}
                >
                  {message}
                </Stack>
              ))}
            </div>

            <Stack
              horizontal
              horizontalAlign="end"
              verticalAlign="end"
              style={{
                display: "flex",
                alignItems: "center",
                borderRadius: "20px",
                background: "white",
                padding: "5px",
                margin: "5px",
              }}
            >
              <TextField
                placeholder="Write your own prompt or ask a question"
                value={userInput}
                onChange={(_, newValue) => handleInputChange(newValue)}
                onKeyDown={handleEnterKeyPress}
                multiline
                resizable={false}
                styles={{
                  root: {
                    width: "100%",
                    height: "80px",
                    borderRadius: "20px",
                    padding: "8px",
                    border: "none",
                    outline: "none",
                  },
                  fieldGroup: { border: "none" },
                }}
              />
              <IconButton iconProps={{ iconName: "Send" }} onClick={handleSendMessage} />
            </Stack>
          </Stack>
        </>
      )}
    </Stack>
  );
};
