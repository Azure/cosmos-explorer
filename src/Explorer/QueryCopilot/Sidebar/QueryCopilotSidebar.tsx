import { IButtonStyles, IconButton, Image, Stack, Text, TextField } from "@fluentui/react";
import { SamplePrompts, SamplePromptsProps } from "Explorer/QueryCopilot/SamplePrompts/SamplePrompts";
import { WelcomeSidebarPopup } from "Explorer/QueryCopilot/Sidebar/WelcomeSidebarPopup";
import { useQueryCopilot } from "hooks/useQueryCopilot";
import React from "react";
import CopilotIcon from "../../../../images/CopilotSidebarLogo.svg";
import HintIcon from "../../../../images/Hint.svg";

const sampleChatMessages: string[] = [
  "Write a query to return last 10 records in the database",
  'Write a query to return all records in this table created in the last thirty days which also have the record owner as "Contoso"',
];

const promptStyles: IButtonStyles = {
  root: { border: "5px", selectors: { ":hover": { outline: "1px dashed #605e5c" } } },
  label: { fontWeight: 400, textAlign: "left", paddingLeft: 8 },
};

export const QueryCopilotSidebar: React.FC = (): JSX.Element => {
  const {
    setWasCopilotUsed,
    showCopilotSidebar,
    setShowCopilotSidebar,
    userPrompt,
    setUserPrompt,
    chatMessages,
    setChatMessages,
    showWelcomeSidebar,
    isSamplePromptsOpen,
    setIsSamplePromptsOpen,
  } = useQueryCopilot();

  const sampleProps: SamplePromptsProps = {
    isSamplePromptsOpen: isSamplePromptsOpen,
    setIsSamplePromptsOpen: setIsSamplePromptsOpen,
    setTextBox: setUserPrompt,
  };

  const handleSendMessage = () => {
    if (userPrompt.trim() !== "") {
      setChatMessages([...chatMessages, userPrompt]);
      setUserPrompt("");
    }
  };

  const handleInputChange = (value: string) => {
    setUserPrompt(value);
  };

  const handleEnterKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter" && !event.shiftKey) {
      event.preventDefault();
      handleSendMessage();
    }
  };

  React.useEffect(() => {
    if (showCopilotSidebar) {
      setWasCopilotUsed(true);
    }
  }, []);

  return (
    <Stack style={{ width: "100%", height: "100%", backgroundColor: "#FAFAFA", overflow: "auto" }}>
      <Stack style={{ margin: "15px 0px 0px 0px", padding: "5px" }}>
        <Stack style={{ display: "flex", justifyContent: "space-between" }} horizontal verticalAlign="center">
          <Stack horizontal verticalAlign="center">
            <Image src={CopilotIcon} />
            <Text style={{ marginLeft: "5px", fontWeight: "bold" }}>Copilot</Text>
            <Text
              style={{
                background: "#f0f0f0",
                fontSize: "10px",
                padding: "2px 4px",
                marginLeft: "5px",
                borderRadius: "8px",
              }}
            >
              Preview
            </Text>
          </Stack>
          <IconButton
            onClick={() => setShowCopilotSidebar(false)}
            iconProps={{ iconName: "Cancel" }}
            title="Exit"
            ariaLabel="Exit"
            style={{ color: "#424242", verticalAlign: "middle" }}
          />
        </Stack>
      </Stack>
      {showWelcomeSidebar ? (
        <Stack.Item styles={{ root: { textAlign: "center", verticalAlign: "middle" } }}>
          <WelcomeSidebarPopup />
        </Stack.Item>
      ) : (
        <>
          <Stack horizontalAlign="center" style={{ color: "#707070" }} tokens={{ padding: 8, childrenGap: 8 }}>
            {new Date().toLocaleDateString("en-US", {
              month: "long",
              day: "numeric",
            })}{" "}
            {new Date().toLocaleTimeString("en-US", {
              hour: "numeric",
              minute: "numeric",
              hour12: true,
            })}
          </Stack>
          <Stack style={{ flexGrow: 1, display: "flex", flexDirection: "column" }}>
            <div style={{ flexGrow: 1, overflowY: "auto" }}>
              <Stack
                tokens={{ padding: 16, childrenGap: 12 }}
                style={{
                  backgroundColor: "white",
                  margin: "5px 10px",
                  borderRadius: "8px",
                }}
              >
                <Text variant="medium">
                  Hello, I am Cosmos Db&apos;s copilot assistant. I can help you do the following things:
                </Text>
                <Stack tokens={{ childrenGap: 8 }}>
                  <Stack horizontal style={{ marginLeft: "15px" }} tokens={{ childrenGap: 8 }}>
                    <Text style={{ fontSize: "16px", lineHeight: "16px", verticalAlign: "middle" }}>•</Text>
                    <Text style={{ verticalAlign: "middle" }}>Generate queries based upon prompt you suggest</Text>
                  </Stack>
                  <Stack horizontal style={{ marginLeft: "15px" }} tokens={{ childrenGap: 8 }}>
                    <Text style={{ fontSize: "16px", lineHeight: "16px", verticalAlign: "middle" }}>•</Text>
                    <Text style={{ verticalAlign: "middle" }}>
                      Explain and provide alternate queries for a query suggested by you
                    </Text>
                  </Stack>
                  <Stack horizontal style={{ marginLeft: "15px" }} tokens={{ childrenGap: 8 }}>
                    <Text style={{ fontSize: "16px", lineHeight: "16px", verticalAlign: "middle" }}>•</Text>
                    <Text style={{ verticalAlign: "middle" }}>Help answer questions about Cosmos DB</Text>
                  </Stack>
                </Stack>
                <Text variant="medium">
                  To get started, ask me a question or use one of the sample prompts to generate a query. AI-generated
                  content may be incorrect.
                </Text>
              </Stack>
              {chatMessages.map((message, index) => (
                <Stack
                  key={index}
                  horizontalAlign="center"
                  tokens={{ padding: 8, childrenGap: 8 }}
                  style={{
                    backgroundColor: "#E0E7FF",
                    borderRadius: "8px",
                    margin: "5px 10px",
                    textAlign: "start",
                  }}
                >
                  {message}
                </Stack>
              ))}
            </div>

            {chatMessages.length === 0 && (
              <Stack
                horizontalAlign="end"
                verticalAlign="end"
                style={{
                  display: "flex",
                  alignItems: "center",
                  padding: "10px",
                  margin: "10px",
                }}
              >
                <Text
                  onClick={() => handleInputChange(sampleChatMessages[0])}
                  style={{
                    cursor: "pointer",
                    border: "1.5px solid #B0BEFF",
                    width: "100%",
                    padding: "2px",
                    borderRadius: "4px",
                    marginBottom: "5px",
                  }}
                >
                  {sampleChatMessages[0]}
                </Text>
                <Text
                  onClick={() => handleInputChange(sampleChatMessages[1])}
                  style={{
                    cursor: "pointer",
                    border: "1.5px solid #B0BEFF",
                    width: "100%",
                    padding: "2px",
                    borderRadius: "4px",
                    marginBottom: "5px",
                  }}
                >
                  {sampleChatMessages[1]}
                </Text>
              </Stack>
            )}

            <Stack
              horizontal
              horizontalAlign="end"
              verticalAlign="end"
              style={{
                display: "flex",
                alignItems: "center",
                borderRadius: "20px",
                background: "white",
                padding: "5px",
                margin: "5px",
              }}
            >
              <Stack>
                <Image src={HintIcon} styles={promptStyles} onClick={() => setIsSamplePromptsOpen(true)} />
                <SamplePrompts sampleProps={sampleProps} />
              </Stack>
              <TextField
                placeholder="Write your own prompt or ask a question"
                value={userPrompt}
                onChange={(_, newValue) => handleInputChange(newValue)}
                onKeyDown={handleEnterKeyPress}
                multiline
                resizable={false}
                styles={{
                  root: {
                    width: "100%",
                    height: "80px",
                    borderRadius: "20px",
                    padding: "8px",
                    border: "none",
                    outline: "none",
                    marginLeft: "10px",
                  },
                  fieldGroup: { border: "none" },
                }}
              />
              <IconButton iconProps={{ iconName: "Send" }} onClick={handleSendMessage} />
            </Stack>
          </Stack>
        </>
      )}
    </Stack>
  );
};
