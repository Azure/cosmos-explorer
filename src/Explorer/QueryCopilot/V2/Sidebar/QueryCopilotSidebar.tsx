import { Stack } from "@fluentui/react";
import { RetrievingBubble } from "Explorer/QueryCopilot/V2/Bubbles/Retriveing/RetrievingBubble";
import { SampleBubble } from "Explorer/QueryCopilot/V2/Bubbles/Sample/SampleBubble";
import { WelcomeBubble } from "Explorer/QueryCopilot/V2/Bubbles/Welcome/WelcomeBubble";
import { Footer } from "Explorer/QueryCopilot/V2/Footer/Footer";
import { Header } from "Explorer/QueryCopilot/V2/Header/Header";
import { useQueryCopilot } from "hooks/useQueryCopilot";
import React from "react";
import { WelcomeSidebarModal } from "../Modal/WelcomeSidebarModal";

export const QueryCopilotSidebar: React.FC = (): JSX.Element => {
  const { setWasCopilotUsed, showCopilotSidebar, chatMessages, showWelcomeSidebar } = useQueryCopilot();

  React.useEffect(() => {
    if (showCopilotSidebar) {
      setWasCopilotUsed(true);
    }
  }, []);

  return (
    <Stack style={{ width: "100%", height: "100%", backgroundColor: "#FAFAFA", overflow: "auto" }}>
      <Header />
      {showWelcomeSidebar ? (
        <WelcomeSidebarModal />
      ) : (
        <>
          <Stack
            style={{
              flexGrow: 1,
              display: "flex",
              flexDirection: "column",
              overflowY: "auto",
            }}
          >
            <WelcomeBubble />
            {chatMessages.map((message, index) => (
              <Stack
                key={index}
                horizontalAlign="center"
                tokens={{ padding: 8, childrenGap: 8 }}
                style={{
                  backgroundColor: "#E0E7FF",
                  borderRadius: "8px",
                  margin: "5px 10px",
                  textAlign: "start",
                }}
              >
                {message}
              </Stack>
            ))}

            <RetrievingBubble />

            {chatMessages.length === 0 && <SampleBubble />}
          </Stack>
          <Footer />
        </>
      )}
    </Stack>
  );
};
