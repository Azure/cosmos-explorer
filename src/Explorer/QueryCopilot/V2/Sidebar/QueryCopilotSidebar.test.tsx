import { Stack } from "@fluentui/react";
import { SampleBubble } from "Explorer/QueryCopilot/V2/Bubbles/Sample/SampleBubble";
import { shallow } from "enzyme";
import { useQueryCopilot } from "hooks/useQueryCopilot";
import { withHooks } from "jest-react-hooks-shallow";
import React from "react";
import { QueryCopilotSidebar } from "./QueryCopilotSidebar";

describe("QueryCopilotSidebar snapshot test", () => {
  const initialState = useQueryCopilot.getState();

  beforeEach(() => {
    useQueryCopilot.setState(initialState, true);
  });

  it("should render and set copilot used flag ", () => {
    withHooks(() => {
      useQueryCopilot.getState().setShowCopilotSidebar(true);
      const wrapper = shallow(<QueryCopilotSidebar />);

      expect(useQueryCopilot.getState().wasCopilotUsed).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("should render and not set copilot used flag ", () => {
    withHooks(() => {
      const wrapper = shallow(<QueryCopilotSidebar />);

      expect(useQueryCopilot.getState().wasCopilotUsed).toBeFalsy();
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("should render with chat messages", () => {
    const message = "some test message";
    useQueryCopilot.getState().setChatMessages([message]);
    const wrapper = shallow(<QueryCopilotSidebar />);

    const messageContainer = wrapper.find(Stack).findWhere((x) => x.text() === message);

    expect(messageContainer).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render samples without messages", () => {
    const wrapper = shallow(<QueryCopilotSidebar />);

    const sampleBubble = wrapper.find(SampleBubble);

    expect(sampleBubble).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
