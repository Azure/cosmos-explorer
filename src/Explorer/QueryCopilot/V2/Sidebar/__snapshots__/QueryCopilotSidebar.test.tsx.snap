// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryCopilotSidebar snapshot test should render and not set copilot used flag  1`] = `
<Stack
  style={
    Object {
      "backgroundColor": "#FAFAFA",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Header />
  <WelcomeSidebarModal />
  <Stack
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "overflowY": "auto",
      }
    }
  >
    <WelcomeBubble />
    <SampleBubble />
  </Stack>
  <Footer />
</Stack>
`;

exports[`Footer snapshot test should render and set copilot used flag  1`] = `
<Stack
  style={
    Object {
      "backgroundColor": "#FAFAFA",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Header />
  <WelcomeSidebarModal />
  <Stack
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "overflowY": "auto",
      }
    }
  >
    <WelcomeBubble />
    <SampleBubble />
  </Stack>
  <Footer />
</Stack>
`;

exports[`Footer snapshot test should render samples without messages 1`] = `
<Stack
  style={
    Object {
      "backgroundColor": "#FAFAFA",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Header />
  <WelcomeSidebarModal />
  <Stack
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "overflowY": "auto",
      }
    }
  >
    <WelcomeBubble />
    <SampleBubble />
  </Stack>
  <Footer />
</Stack>
`;

exports[`Footer snapshot test should render with chat messages 1`] = `
<Stack
  style={
    Object {
      "backgroundColor": "#FAFAFA",
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Header />
  <WelcomeSidebarModal />
  <Stack
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "flexGrow": 1,
        "overflowY": "auto",
      }
    }
  >
    <WelcomeBubble />
    <Stack
      horizontalAlign="center"
      key="0"
      style={
        Object {
          "backgroundColor": "#E0E7FF",
          "borderRadius": "8px",
          "margin": "5px 10px",
          "textAlign": "start",
        }
      }
      tokens={
        Object {
          "childrenGap": 8,
          "padding": 8,
        }
      }
    >
      some test message
    </Stack>
  </Stack>
  <Footer />
</Stack>
`;
